pkg/server/api/trustdomain/v1/service_test.go

```diff
@@ -1 +1,579 @@
+						Message: "unable to create federation relationship for server trust domain",
+					},
+				},
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ds := newFakeDS(t)
+			ds.customDSResponse = tt.customDSReponse
+
+			test := setupServiceTest(t, ds)
+			defer test.Cleanup()
+
+			// Create default relationship
+			createTestRelationships(t, ds, defaultFederationRelationship)
+
+			// Setup fake
+			ds.AppendNextError(tt.expectDSErr)
+
+			// Batch create
+			resp, err := test.client.BatchCreateFederationRelationship(ctx, &trustdomainv1.BatchCreateFederationRelationshipRequest{
+				FederationRelationship: tt.req,
+				OutputMask:             tt.outputMask,
+			})
+
+			require.NoError(t, err)
+			require.NotNil(t, resp)
+			spiretest.AssertLogs(t, test.logHook.AllEntries(), tt.expectLogs)
+
+			spiretest.AssertProtoEqual(t, &trustdomainv1.BatchCreateFederationRelationshipResponse{
+				Results: tt.expectResults,
+			}, resp)
+		})
+	}
+}
+
+func createTestRelationships(t *testing.T, ds datastore.DataStore, relationships ...*datastore.FederationRelationship) map[string]*datastore.FederationRelationship {
+	frMap := make(map[string]*datastore.FederationRelationship)
+
+	for _, fr := range relationships {
+		dRelationship, err := ds.CreateFederationRelationship(ctx, fr)
+		require.NoError(t, err)
+
+		frMap[dRelationship.TrustDomain.String()] = fr
+	}
+
+	return frMap
+}
+
+type serviceTest struct {
+	client  trustdomainv1.TrustDomainClient
+	ds      datastore.DataStore
+	logHook *test.Hook
+	done    func()
+}
+
+func (s *serviceTest) Cleanup() {
+	s.done()
+}
+
+func setupServiceTest(t *testing.T, ds datastore.DataStore) *serviceTest {
+	service := trustdomain.New(trustdomain.Config{
+		DataStore:   ds,
+		TrustDomain: td,
+	})
+
+	log, logHook := test.NewNullLogger()
+	log.Level = logrus.DebugLevel
+	registerFn := func(s *grpc.Server) {
+		trustdomain.RegisterService(s, service)
+	}
+
+	test := &serviceTest{
+		ds:      ds,
+		logHook: logHook,
+	}
+
+	ppMiddleware := middleware.Preprocess(func(ctx context.Context, fullMethod string, req interface{}) (context.Context, error) {
+		ctx = rpccontext.WithLogger(ctx, log)
+
+		return ctx, nil
+	})
+
+	unaryInterceptor, streamInterceptor := middleware.Interceptors(middleware.Chain(
+		ppMiddleware,
+		// Add audit log with uds tracking disabled
+		middleware.WithAuditLog(false),
+	))
+
+	server := grpc.NewServer(
+		grpc.UnaryInterceptor(unaryInterceptor),
+		grpc.StreamInterceptor(streamInterceptor),
+	)
+
+	conn, done := spiretest.NewAPIServerWithMiddleware(t, registerFn, server)
+	test.done = done
+	test.client = trustdomainv1.NewTrustDomainClient(conn)
+
+	return test
+}
+
+type fakeDS struct {
+	*fakedatastore.DataStore
+
+	customDSResponse *datastore.FederationRelationship
+}
+
+func newFakeDS(t *testing.T) *fakeDS {
+	return &fakeDS{
+		DataStore: fakedatastore.New(t),
+	}
+}
+
+// CreateFederationRelationship create custom function to validate custom relationship returned from ds
```

Comment:
I don't think this comment is too useful to have on the fake, but if you do want to keep it, this comment seems misleading since this function does not create a function nor do any validation?
```suggestion
// CreateFederationRelationship wraps the datastore method of the same name and provides an injection hook for a custom response.
```